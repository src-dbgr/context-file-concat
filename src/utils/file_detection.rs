use anyhow::Result;
use std::fs::File;
use std::io::{BufReader, Read};
use std::path::Path;
use std::sync::OnceLock;

const TEXT_EXTENSIONS: &[&str] = &[
    "txt",
    "md",
    "markdown",
    "rst",
    "asciidoc",
    "adoc",
    "rs",
    "py",
    "js",
    "ts",
    "jsx",
    "tsx",
    "java",
    "c",
    "cpp",
    "cxx",
    "cc",
    "h",
    "hpp",
    "hxx",
    "go",
    "rb",
    "php",
    "swift",
    "kt",
    "kts",
    "scala",
    "clj",
    "cljs",
    "hs",
    "ml",
    "fs",
    "fsx",
    "html",
    "htm",
    "xml",
    "xhtml",
    "css",
    "scss",
    "sass",
    "less",
    "svg",
    "vue",
    "svelte",
    "json",
    "yaml",
    "yml",
    "toml",
    "ini",
    "cfg",
    "conf",
    "config",
    "properties",
    "sql",
    "sh",
    "bash",
    "zsh",
    "fish",
    "ps1",
    "bat",
    "cmd",
    "dockerfile",
    "makefile",
    "cmake",
    "gradle",
    "maven",
    "pom",
    "build",
    "tex",
    "bib",
    "r",
    "m",
    "pl",
    "lua",
    "vim",
    "el",
    "lisp",
    "dart",
    "elm",
    "ex",
    "exs",
    "erl",
    "hrl",
    "nim",
    "crystal",
    "cr",
    "zig",
    "odin",
    "v",
    "log",
    "trace",
    "out",
    "err",
    "diff",
    "patch",
    "gitignore",
    "gitattributes",
    "editorconfig",
    "env",
    "example",
    "sample",
    "template",
    "spec",
    "test",
    "readme",
    "license",
    "changelog",
    "todo",
    "notes",
    "doc",
    "docs",
    "man",
    "help",
    "faq",
    "lock",
    "sum",
    "mod",
    "work",
    "pest",
    "ron",
    "map",
    "d.ts",
    "mjs",
    "cjs",
    "coffee",
    "litcoffee",
    "ls",
    "flow",
    "pegjs",
    "graphql",
    "gql",
    "prisma",
    "proto",
    "thrift",
    "avsc",
    "jsonl",
    "ndjson",
    "csv",
    "tsv",
    "psv",
    "ssv",
    "tab",
    "data",
    "idx",
    "org",
    "cls",
    "sty",
    "bst",
    "aux",
    "fdb_latexmk",
    "fls",
    "R",
    "Rmd",
    "Rnw",
    "jl",
    "ipynb",
    "pyx",
    "pxd",
    "pxi",
    "pyi",
    "gnumakefile",
    "containerfile",
    "vagrantfile",
    "rakefile",
    "gemfile",
    "guardfile",
    "procfile",
    "capfile",
    "berksfile",
    "jenkinsfile",
    "dangerfile",
    "fastfile",
    "appfile",
    "deliverfile",
    "snapfile",
    "ignore",
    "keep",
    "gitkeep",
    "npmignore",
    "dockerignore",
    "eslintrc",
    "babelrc",
    "browserslistrc",
    "nvmrc",
    "rvmrc",
    "rbenv-version",
    "ruby-version",
    "node-version",
    "wasm",
    "wat",
    "wit",
    "component",
    "lalrpop",
    "y",
    "l",
    "lex",
    "yacc",
    "capnp",
    "fbs",
    "schema",
    "avdl",
    "gn",
    "gni",
    "bp",
    "BUILD",
    "WORKSPACE",
    "bzl",
    "nix",
    "drv",
    "store-path",
    "dhall",
    "purescript",
    "purs",
    "roc",
    "gleam",
    "grain",
    "hx",
    "hxml",
    "moon",
    "just",
    "justfile",
    "task",
    "taskfile",
    "clang-format",
    "rustfmt",
    "modulemap",
    "def",
    "exports",
    "version",
    "in",
    "am",
    "ac",
    "m4",
    "ctest",
    "service",
    "socket",
    "timer",
    "mount",
    "desktop",
    "appdata",
    "metainfo",
];

const IMAGE_EXTENSIONS: &[&str] = &[
    "jpg", "jpeg", "png", "gif", "bmp", "ico", "webp", "tiff", "tif", "raw", "cr2", "nef", "orf",
    "dng", "heic", "heif", "avif", "jfif",
];

const BINARY_EXTENSIONS: &[&str] = &[
    "exe",
    "dll",
    "so",
    "dylib",
    "app",
    "deb",
    "rpm",
    "msi",
    "zip",
    "tar",
    "gz",
    "bz2",
    "7z",
    "rar",
    "jar",
    "war",
    "mp3",
    "mp4",
    "avi",
    "mkv",
    "mov",
    "wmv",
    "flv",
    "webm",
    "pdf",
    "docx",
    "xlsx",
    "pptx",
    "bin",
    "dat",
    "db",
    "sqlite",
    "sqlite3",
    "rlib",
    "rmeta",
    "d",
    "pdb",
    "ilk",
    "exp",
    "lib",
    "a",
    "obj",
    "o",
    "class",
    "pyc",
    "pyo",
    "__pycache__",
    "cache",
    "tmp",
    "temp",
    "swap",
    "bak",
    "backup",
    "fingerprint",
    "deps",
    "incremental",
    "crate",
    "gem",
    "whl",
    "egg",
    "snap",
    "flatpak",
];

static TEXT_EXT_SET: OnceLock<std::collections::HashSet<&'static str>> = OnceLock::new();

static BINARY_EXT_SET: OnceLock<std::collections::HashSet<&'static str>> = OnceLock::new();

static IMAGE_EXT_SET: OnceLock<std::collections::HashSet<&'static str>> = OnceLock::new();

fn get_text_ext_set() -> &'static std::collections::HashSet<&'static str> {
    TEXT_EXT_SET.get_or_init(|| TEXT_EXTENSIONS.iter().copied().collect())
}

fn get_binary_ext_set() -> &'static std::collections::HashSet<&'static str> {
    BINARY_EXT_SET.get_or_init(|| BINARY_EXTENSIONS.iter().copied().collect())
}

fn get_image_ext_set() -> &'static std::collections::HashSet<&'static str> {
    IMAGE_EXT_SET.get_or_init(|| IMAGE_EXTENSIONS.iter().copied().collect())
}

const MAX_CONTENT_CHECK_SIZE: u64 = 20 * 1024 * 1024;

const CONTENT_CHECK_BUFFER_SIZE: usize = 1024;

/// Determines if a file is likely to be a text file.
pub fn is_text_file(path: &Path) -> Result<bool> {
    if let Some(extension) = path.extension().and_then(|ext| ext.to_str()) {
        let ext_lower = extension.to_lowercase();
        if get_text_ext_set().contains(ext_lower.as_str()) {
            return Ok(true);
        }
        if get_binary_ext_set().contains(ext_lower.as_str()) {
            return Ok(false);
        }
        if get_image_ext_set().contains(ext_lower.as_str()) {
            return Ok(false);
        }
    }
    match std::fs::metadata(path) {
        Ok(metadata) => {
            if metadata.len() > MAX_CONTENT_CHECK_SIZE {
                return Ok(false);
            }
            if metadata.len() == 0 {
                return Ok(true);
            }
        }
        Err(_) => return Ok(false),
    }
    check_file_content_optimized(path)
}

/// Determines if a file is an image file.
#[allow(dead_code)]
pub fn is_image_file(path: &Path) -> bool {
    if let Some(extension) = path.extension().and_then(|ext| ext.to_str()) {
        let ext_lower = extension.to_lowercase();
        get_image_ext_set().contains(ext_lower.as_str()) || ext_lower == "svg"
    } else {
        false
    }
}

/// Checks the initial bytes of a file to guess if it's text.
fn check_file_content_optimized(path: &Path) -> Result<bool> {
    let mut buffer = [0u8; CONTENT_CHECK_BUFFER_SIZE];
    let file = File::open(path)?;
    let mut reader = BufReader::with_capacity(CONTENT_CHECK_BUFFER_SIZE, file);
    let bytes_read = reader.read(&mut buffer)?;
    if bytes_read == 0 {
        return Ok(true);
    }
    if buffer[..bytes_read].contains(&0) {
        return Ok(false);
    }
    match std::str::from_utf8(&buffer[..bytes_read]) {
        Ok(_) => Ok(true),
        Err(_) => {
            // VET: Switched to a more robust heuristic. Instead of checking for printable
            // ASCII, we check for the absence of control characters (excluding whitespace).
            // This correctly handles legacy 8-bit encodings like Latin-1.
            let control_char_count = buffer[..bytes_read]
                .iter()
                .filter(|&&b| b < 32 && b != 9 && b != 10 && b != 13)
                .count();

            // If less than 5% of the bytes are weird control characters, we assume it's text.
            let ratio = control_char_count as f32 / bytes_read as f32;
            Ok(ratio < 0.05)
        }
    }
}

/// Classifies a batch of files as text or image.
#[allow(dead_code)]
pub fn batch_classify_files(paths: &[&Path]) -> Vec<(bool, bool)> {
    paths
        .iter()
        .map(|path| {
            let is_text = is_text_file(path).unwrap_or(false);
            let is_image = if !is_text { is_image_file(path) } else { false };
            (is_text, is_image)
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;
    use tempfile::TempDir;

    /// Helper to create a test file with specific content in a temporary directory.
    fn create_test_file(dir: &TempDir, name: &str, content: &[u8]) -> std::path::PathBuf {
        let path = dir.path().join(name);
        fs::write(&path, content).expect("Failed to write test file");
        path
    }

    #[test]
    fn test_is_text_by_known_text_extension() {
        let dir = TempDir::new().unwrap();
        let text_files = [
            "document.txt",
            "code.rs",
            "script.py",
            "config.YAML",
            "Makefile",
        ];
        for name in text_files {
            let path = create_test_file(&dir, name, b"some content");
            assert!(
                is_text_file(&path).unwrap(),
                "Expected '{}' to be a text file",
                name
            );
        }
    }

    #[test]
    fn test_is_not_text_by_known_binary_extension() {
        let dir = TempDir::new().unwrap();
        let bin_files = ["archive.zip", "program.EXE", "library.dll", "document.pdf"];
        for name in bin_files {
            let path = create_test_file(&dir, name, b"\xDE\xAD\xBE\xEF");
            assert!(
                !is_text_file(&path).unwrap(),
                "Expected '{}' to be a binary file",
                name
            );
        }
    }

    #[test]
    fn test_is_not_text_by_known_image_extension() {
        let dir = TempDir::new().unwrap();
        let img_files = ["photo.jpg", "logo.PNG", "animated.gif"];
        for name in img_files {
            let path = create_test_file(&dir, name, b"imagedata");
            assert!(
                !is_text_file(&path).unwrap(),
                "Expected '{}' to be a binary (image) file",
                name
            );
        }
    }

    #[test]
    fn test_nonexistent_file_is_not_text() {
        let dir = TempDir::new().unwrap();
        let path = dir.path().join("nonexistent.file");
        assert!(!is_text_file(&path).unwrap());
    }

    #[test]
    fn test_empty_file_is_text() {
        let dir = TempDir::new().unwrap();
        let path = create_test_file(&dir, "empty.unknown", b"");
        assert!(is_text_file(&path).unwrap());
    }

    #[test]
    fn test_is_text_by_utf8_content() {
        let dir = TempDir::new().unwrap();
        let path = create_test_file(
            &dir,
            "content.unknown",
            "Hello, world! This is UTF-8. 🦀".as_bytes(),
        );
        assert!(is_text_file(&path).unwrap());
    }

    #[test]
    fn test_is_not_text_due_to_null_byte() {
        let dir = TempDir::new().unwrap();
        let path = create_test_file(
            &dir,
            "binary.unknown",
            b"Here is some text\x00with a null byte.",
        );
        assert!(!is_text_file(&path).unwrap());
    }

    #[test]
    fn test_is_text_by_printable_ratio_heuristic() {
        let dir = TempDir::new().unwrap();
        let latin1_text = b"K\xf6nnen"; // "Können" in Latin-1
        let path = create_test_file(&dir, "legacy_encoding.unknown", latin1_text);
        assert!(is_text_file(&path).unwrap());
    }

    #[test]
    fn test_is_not_text_due_to_low_printable_ratio() {
        let dir = TempDir::new().unwrap();
        let random_binary = b"abc\x01\x02\x03\x04\x05\x06\x07\x08\x80\x90\xA0";
        let path = create_test_file(&dir, "random.unknown", random_binary);
        assert!(!is_text_file(&path).unwrap());
    }

    #[test]
    fn test_is_image_file_logic() {
        assert!(is_image_file(&Path::new("image.jpg")));
        assert!(is_image_file(&Path::new("image.PNG")));
        assert!(is_image_file(&Path::new("image.svg")));
        assert!(!is_image_file(&Path::new("document.txt")));
        assert!(!is_image_file(&Path::new("archive.zip")));
        assert!(!is_image_file(&Path::new("no_extension")));
    }

    #[test]
    fn test_batch_classify_files_logic() {
        let dir = TempDir::new().unwrap();
        let p1 = create_test_file(&dir, "text.txt", b"text");
        let p2 = create_test_file(&dir, "image.png", b"png");
        let p3 = create_test_file(&dir, "binary.zip", b"zip");
        let p4 = create_test_file(&dir, "vector.svg", b"<svg>");

        let paths: Vec<&Path> = vec![&p1, &p2, &p3, &p4];
        let results = batch_classify_files(&paths);

        assert_eq!(results.len(), 4);
        // (is_text, is_image)
        assert_eq!(results[0], (true, false));
        assert_eq!(results[1], (false, true));
        assert_eq!(results[2], (false, false));
        assert_eq!(results[3], (true, false));
    }
}
