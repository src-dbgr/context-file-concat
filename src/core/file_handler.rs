use std::path::{Path, PathBuf};
use std::fs;
use std::io::{BufRead, BufReader};
use anyhow::Result;
use tokio::sync::mpsc;

use super::ScanProgress;
use crate::utils::file_detection::is_text_file;

pub struct FileHandler;

impl FileHandler {
    // MODIFIED: This function now generates content in-memory and returns it, instead of writing to a file.
    pub async fn generate_concatenated_content(
        selected_files: &[PathBuf],
        include_tree: bool,
        tree_content: Option<String>,
        progress_sender: mpsc::UnboundedSender<ScanProgress>,
    // MODIFIED: Returns the content string, its size, and line count
    ) -> Result<(String, u64, usize)> {
        let mut content = String::new();
        let total_files = selected_files.len();
        
        // Add header
        content.push_str("# ContextFileConcat Output\n");
        content.push_str(&format!("# Generated: {}\n", chrono::Local::now().format("%Y-%m-%d %H:%M:%S")));
        content.push_str(&format!("# Total files: {}\n", total_files));
        content.push_str("\n");
        
        // Process each file
        for (i, file_path) in selected_files.iter().enumerate() {
            progress_sender.send(ScanProgress {
                current_file: file_path.clone(),
                processed: i,
                total: total_files,
                status: format!("Processing file {}/{}", i + 1, total_files),
                file_size: None,
                line_count: None,
            })?;
            
            // Skip directories
            if file_path.is_dir() {
                continue;
            }
            
            // Check if file is text
            if !is_text_file(file_path).unwrap_or(false) {
                content.push_str(&format!("{}\n", file_path.display()));
                content.push_str("----------------------------------------------------\n");
                content.push_str("[BINARY FILE - CONTENT SKIPPED]\n");
                content.push_str("----------------------------------------------------\n\n");
                continue;
            }
            
            // Add file header
            content.push_str(&format!("{}\n", file_path.display()));
            content.push_str("----------------------------------------------------\n");
            
            // Read and add file content
            match Self::read_file_content(file_path) {
                Ok(file_content) => {
                    content.push_str(&file_content);
                    if !file_content.ends_with('\n') {
                        content.push('\n');
                    }
                }
                Err(e) => {
                    content.push_str(&format!("[ERROR READING FILE: {}]\n", e));
                }
            }
            
            content.push_str("----------------------------------------------------\n\n");
            
            // Yield control periodically
            if i % 10 == 0 {
                tokio::task::yield_now().await;
            }
        }
        
        // Add tree at the end if requested
        if include_tree {
            if let Some(tree) = tree_content {
                content.push_str("\n# DIRECTORY TREE\n");
                content.push_str("====================================================\n");
                content.push_str(&tree);
                content.push_str("====================================================\n");
            }
        }
        
        // Final progress update before returning
        progress_sender.send(ScanProgress {
            current_file: PathBuf::from("Finalizing..."),
            processed: total_files,
            total: total_files,
            status: "Finalizing content...".to_string(),
            file_size: None,
            line_count: None,
        })?;
        
        // Calculate file statistics from the generated string
        let file_size = content.len() as u64;
        let line_count = content.lines().count();

        progress_sender.send(ScanProgress {
            current_file: PathBuf::from("Generated Preview"),
            processed: total_files,
            total: total_files,
            status: "Complete!".to_string(),
            file_size: Some(file_size),
            line_count: Some(line_count),
        })?;
        
        tracing::info!("Successfully generated concatenated content in memory.");
        Ok((content, file_size, line_count))
    }
    
    fn read_file_content(file_path: &Path) -> Result<String> {
        // Check file size first (20MB limit)
        let metadata = fs::metadata(file_path)?;
        if metadata.len() > 20 * 1024 * 1024 {
            return Ok(format!("[FILE TOO LARGE: {} bytes - CONTENT SKIPPED]", metadata.len()));
        }
        
        // Try to read as UTF-8 first
        match fs::read_to_string(file_path) {
            Ok(content) => Ok(content),
            Err(_) => {
                // If UTF-8 fails, try reading as bytes and converting
                let bytes = fs::read(file_path)?;
                match String::from_utf8_lossy(&bytes) {
                    content if content.contains('\u{FFFD}') => {
                        Ok("[BINARY OR NON-UTF8 FILE - CONTENT SKIPPED]".to_string())
                    }
                    content => Ok(content.to_string()),
                }
            }
        }
    }
    
    pub fn get_file_preview(file_path: &Path, max_lines: usize) -> Result<String> {
        if file_path.is_dir() {
            return Ok("[DIRECTORY]".to_string());
        }
        
        if !is_text_file(file_path).unwrap_or(false) {
            return Ok("[BINARY FILE]".to_string());
        }
        
        let file = fs::File::open(file_path)?;
        let reader = BufReader::new(file);
        let mut preview = String::new();
        let mut line_count = 0;
        
        for line in reader.lines() {
            if line_count >= max_lines {
                preview.push_str("...\n[Preview truncated]");
                break;
            }
            
            match line {
                Ok(line_content) => {
                    preview.push_str(&line_content);
                    preview.push('\n');
                }
                Err(_) => {
                    preview.push_str("[ERROR READING LINE]\n");
                }
            }
            
            line_count += 1;
        }
        
        Ok(preview)
    }

}