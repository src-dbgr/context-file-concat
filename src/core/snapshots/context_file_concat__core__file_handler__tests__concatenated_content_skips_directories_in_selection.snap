---
source: src/core/file_handler.rs
expression: content
---
# CFC Output - Generated: [DATETIME]
# Total files: 2

[PROJECT_NAME]/src/main.rs
===FILE-START===
fn main() {
    println!("Hello, world!");
}
---FILE-END-----
